repository/dimensions_repository.go

```diff
@@ -0,0 +1,130 @@
+package repository
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/ONSdigital/dp-dimension-importer/client"
+	logKeys "github.com/ONSdigital/dp-dimension-importer/common"
+	"github.com/ONSdigital/dp-dimension-importer/model"
+	"github.com/ONSdigital/go-ns/log"
+	bolt "github.com/johnnadratowski/golang-neo4j-bolt-driver"
+)
+
+const (
+	// Create a unique constraint on the dimension type value.
+	uniqueDimConstStmt = "CREATE CONSTRAINT ON (d:`%s`) ASSERT d.value IS UNIQUE"
+
+	// Create the dimension node and the HAS_DIMENSION relationship to the Instance it belongs to.
+	createDimensionAndInstanceRelStmt = "MATCH (i:`%s`) CREATE (d:`%s` {value: {value}}) CREATE (i)-[:HAS_DIMENSION]->(d) RETURN ID(d)"
+
+	stmtKey                   = "statment"
+	stmtParamsKey             = "params"
+	valueKey                  = "value"
+	dimensionsKey             = "dimensions"
+	dimensionsList            = "dimensions_list"
+	uniqueConstErr            = "Error executing unique constraint Statment"
+	uniqueConstSuccess        = "Successfully created unique constraint on dimension"
+	dimensionNilErr           = "Dimension is required but was nil"
+	dimensionInvalidErr       = "Dimension invalid: both dimension.dimension_id and dimension.value are required but were both empty"
+	dimensionIDRequiredErr    = "dimension.Dimension_ID is required but was empty"
+	dimensionValueRequiredErr = "dimension.Value is required but was empty"
+	nodeIDCastErr             = "Unexpected error while casting NodeID to int64"
+	insertDimValidationMsg    = "InsertDimension: dimension failed validation"
+	errExecutingStatment      = "Error executing statement"
+	errReturningRows          = "Error return query rows"
+)
+
+type NeoDatabase interface {
+	Query(query string, params map[string]interface{}) (*client.NeoRows, error)
+	ExecStmt(query string, params map[string]interface{}) (bolt.Result, error)
+}
+
+type DimensionRepository struct {
+	ConstraintsCache map[string]string
```

Comment:
This object only lives for the duration of processing the instance. A new one is created for each instance to keep track of the dimensions its already created constraints for. Once its complete it will die.